src/os/bluestore/bluestore_types.h

```diff
@@ -156,7 +156,8 @@ struct denc_traits<vector<bluestore_pextent_t>> {
   static void bound_encode(const vector<bluestore_pextent_t>& v, size_t& p) {
     p += sizeof(uint32_t);
     size_t per = 0;
-    denc(*(bluestore_pextent_t*)nullptr, per);
+    bluestore_pextent_t tmp;
```

Comment:
@lixiaoy1 the segfault is due to the unoptimized code generated by the compiler, mostly likely you were testing an osd compiled with `-O0`. that's why the codec of `bluestore_pextent_t` is trying to dereference nullptr.

as you can see, `denc.h` is still using nullptr to calculate the size of its element,
```c++
  template<typename U=T>
  static typename std::enable_if<sizeof(U) &&
                                 traits::bounded &&
                                 !traits::featured, void>::type
  bound_encode(const std::vector<T>& s, size_t& p) {
    size_t elem_size = 0;
    denc(*(const T*)nullptr, elem_size);
    p += sizeof(uint32_t) + elem_size * s.size();
  }
```

i don't see a simple way at this moment to fix this other than to use `-O2` to urge the compiler to optimize out the dereference code.

@adamemerson what do you think?